examples/edwards25519-ecvrf/src/edwards25519-ecvrf.rs

```diff
@@ -0,0 +1,438 @@
+    for _ctr in 1..256 {
+        if h.clone() == Option::<EdPoint>::None {
+            let ctr_string = x.to_byte_seq_be().slice(31,1);
+            let hash_string = sha512(&SUITE_STRING
+                .concat(&ONE)
+                .concat(encode_to_curve_salt)
+                .concat(alpha)
+                .concat(&ctr_string)
+                .concat(&ZERO));
+            h = decompress(CompressedEdPoint::from_slice(&hash_string, 0, 32));
+            x = x + Ed25519FieldElement::ONE();
+        }
+    }
+    let h = h.ok_or(Errorec::FailedE2C)?;
+    Ok(point_mul_by_cofactor(h))
+}
+
+// See section 5.4.1.2 (ciphersuite 4) for the ELL2 method:
+// https://datatracker.ietf.org/doc/draft-irtf-cfrg-vrf/
+// Uses the edwards25519-hash crate to convert an arbitrary string to an
+// elliptic curve point
+fn ecvrf_encode_to_curve_h2c_suite(
+    encode_to_curve_salt: &ByteSeq, alpha: &ByteSeq
+) -> EdPointResult {
+    let string_to_be_hashed = encode_to_curve_salt.concat(alpha);
+    let dst = H2C_SUITE_ID_STRING.concat(&SUITE_STRING);
+    let h = ed_encode_to_curve(&string_to_be_hashed, &dst);
+    let h = h.ok().ok_or(Errorec::FailedE2C)?;
+    Ok(h)
+}
+
+// See section 5.4.2 for nonce generation:
+// https://datatracker.ietf.org/doc/draft-irtf-cfrg-vrf/
+// This implements section 5.4.2.2 that is also based on RFC8032
+fn ecvrf_nonce_generation(
+    sk: SecretKey, h_string: &ByteSeq
+) -> Scalar {
+    let hashed_sk_string = sha512(&sk.to_le_bytes());
+    let truncated_hashed_sk_string = hashed_sk_string.slice(32,32);
+    let k_string = sha512(&truncated_hashed_sk_string.concat(h_string));
+    
+    let nonce = BigScalar::from_byte_seq_le(k_string);
+    let nonceseq = nonce.to_byte_seq_le().slice(0, 32);
+    Scalar::from_byte_seq_le(nonceseq)
+}
+
+// See section 5.4.3 for challenge generation
+// https://datatracker.ietf.org/doc/draft-irtf-cfrg-vrf/
+fn ecvrf_challenge_generation(
+    p1: EdPoint, p2: EdPoint, p3: EdPoint, p4: EdPoint, p5: EdPoint
+) -> Scalar {
+    let string = SUITE_STRING
+        .concat(&TWO)
+        .concat(&encode(p1))
+        .concat(&encode(p2))
+        .concat(&encode(p3))
+        .concat(&encode(p4))
+        .concat(&encode(p5))
+        .concat(&ZERO);
+
+    let c_string = sha512(&string);
+    let truncated_c_string= c_string.slice(0, C_LEN).concat(&ByteSeq::new(16));    
+    Scalar::from_byte_seq_le(truncated_c_string)
+}
+
+// See section 5.4.4 for decode proof
+// https://datatracker.ietf.org/doc/draft-irtf-cfrg-vrf/
+fn ecvrf_decode_proof(pi: &ByteSeq) -> ProofResult {
+    let gamma_string = pi.slice(0, PT_LEN);
+    let c_string = pi.slice(PT_LEN, C_LEN);
+    let s_string = pi.slice(PT_LEN + C_LEN, Q_LEN);
+    let gamma = decompress(CompressedEdPoint::from_slice(&gamma_string, 0, 32))
+                .ok_or(Errorec::InvalidProof)?;
+
+    let c = Scalar::from_byte_seq_le(c_string.concat(&ByteSeq::new(16)));
+    let s = Scalar::from_byte_seq_le(s_string.clone());
+
```

Comment:
This looks slightly different than in the standard, where s_test is missing. 
Malthe says it is OK.
> The s_test variable that is used to check that s is greater than the group order. 
When s is created it is automatically mod q, so it does not make sense to check for s. 
s_test on the other hand uses the field LargeMod where the order is as high as can be, and therefore it is possible to actually compare against the group order q.

This seems correct to me. @franziskuskiefer would there be a more idiomatic way of writing this?

Reply:
Yeah I think this is the best we can do right now.
Ideally, `Scalar::from_byte_seq_le` would do this check and return an error. But I don't think it does at the moment.